File: Assignments/Recursion/Bin_search.java
import java.util.Scanner;

public class Bin_search {
    public static void main(String[] args) {
        // Create Scanner object for user input
        Scanner in = new Scanner(System.in);

        // Input length of the array
        System.out.print("Enter the length of the array: ");
        int n = in.nextInt();

        // Input the array elements
        System.out.println("Enter the SORTED array: ");
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = in.nextInt();
        }

        // Input the search element
        System.out.print("Enter the search element: ");
        int ele = in.nextInt();

        // Perform Binary Search 
        int s =0;
        int e = arr.length -1;
        System.out.println("The index of " + ele + " is " + search(arr, ele, s, e));

        // Close Scanner object to prevent resource leak
        in.close();
    }

    //Function definition for Binary Search using recursion
    //Taking the start and end as arguments as we need them everytime we have to call the function (recursive call)
    public static int search(int[] arr, int target, int start, int end) {
        //Start must always be smaller than the end for binary search
        if (start > end) {
            return -1;
        }
        
        //Calculating the value on mid
        int mid = start + (end - start) / 2;
        
        //target found at index mid
        if(target == arr[mid]){
            return mid;
        }
        if (target > arr[mid]) {
            return search(arr, target, mid + 1, end);   //recursive call
        }

        //if(target < arr[mid])
        return search(arr, target, start, mid-1);   //recursive function call
    }
}

File: DSA_Archive/Bin_Search.cpp
#include <iostream>
#include <vector>

using namespace std;

int bs(vector<int>& arr, int s, int e, int t) {
    if (s <= e) {
        int mid = s + (e - s) / 2;

        if (arr[mid] == t) {
            return mid;
        } 
        else if (arr[mid] > t) {
            return bs(arr, s, mid - 1, t);
        } 
        else {
            return bs(arr, mid + 1, e, t);
        }
    }
    return -1; // Not found
}

int main() {
    int n, t;
    cout << "Enter the size of the array: ";
    cin >> n;  // Read size first

    vector<int> arr(n);  // Now declare the vector with the correct size

    cout << "Enter the elements: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i]; // Input elements
    }

    cout << "Enter the target element: ";
    cin >> t;

    int result = bs(arr, 0, n - 1, t);
    
    if (result != -1)
        cout << "Element found at index " << result << endl;
    else
        cout << "Element not found" << endl;

    return 0;
}
